api/src/utils/common-challenge-functions.ts

```diff
@@ -0,0 +1,214 @@
+        file =>
+          pick(file, [
+            'contents',
+            'key',
+            'index',
+            'name',
+            'path',
+            'ext'
+          ]) as CompletedChallengeFile
+      )
+    };
+  } else {
+    completedChallenge = omit(_completedChallenge, ['files']);
+  }
+  let finalChallenge = {} as CompletedChallenge;
+
+  // Since these values are destuctured for easier updating, collectively update before returning
+  const {
+    timezone: userTimezone,
+    completedChallenges = [],
+    needsModeration = false,
+    savedChallenges = [],
+    progressTimestamps = [],
+    partiallyCompletedChallenges = []
+  } = user;
+
+  const userCompletedChallenges: CompletedChallenge[] = completedChallenges;
+  const userSavedChallenges: SavedChallenge[] = savedChallenges;
+  const userProgressTimestamps = progressTimestamps;
+  const userPartiallyCompletedChallenges = partiallyCompletedChallenges;
+
+  const oldIndex = userCompletedChallenges.findIndex(
+    ({ id }) => challengeId === id
+  );
+
+  const alreadyCompleted = oldIndex !== -1;
+  const oldChallenge = alreadyCompleted
+    ? userCompletedChallenges[oldIndex]
+    : null;
+
+  if (alreadyCompleted && oldChallenge) {
+    const challengeWithOldDate = userCompletedChallenges[oldIndex];
+
+    if (challengeWithOldDate) {
+      challengeWithOldDate.completedDate = oldChallenge.completedDate;
+      finalChallenge = {
+        ...completedChallenge,
+        ...challengeWithOldDate
+      };
+    }
+  } else {
+    finalChallenge = {
+      ...completedChallenge
+    };
+    if (userProgressTimestamps && Array.isArray(userProgressTimestamps)) {
+      userProgressTimestamps.push(newProgressTimeStamp);
+    }
+    userCompletedChallenges.push(finalChallenge);
+  }
+
+  if (savableChallenges.includes(challengeId)) {
+    const challengeToSave: SavedChallenge = {
+      id: challengeId,
+      lastSavedDate: newProgressTimeStamp,
+      files: files?.map(file =>
+        pick(file, ['contents', 'key', 'name', 'ext', 'history'])
+      ) as SavedChallengeFile[]
+    };
+
+    const savedIndex = userSavedChallenges.findIndex(
+      ({ id }) => challengeId === id
+    );
+
+    if (savedIndex >= 0) {
+      userSavedChallenges[savedIndex] = challengeToSave;
+    } else {
+      userSavedChallenges.push(challengeToSave);
+    }
+  }
+
+  // remove from partiallyCompleted on submit
+  const updatedPartiallyCompletedChallenges =
+    userPartiallyCompletedChallenges.filter(
+      challenge => challenge.id !== challengeId
+    );
+
+  if (
+    timezone &&
+    timezone !== 'UTC' &&
+    !userTimezone &&
+    userTimezone === 'UTC'
+  ) {
+    timezone = userTimezone;
+    await fastify.prisma.user.update({
+      where: { id: user.id },
+      data: {
+        timezone
+      }
+    });
+  }
+
+  if (needsModeration) {
+    await fastify.prisma.user.update({
+      where: { id: user.id },
+      data: {
+        needsModeration: true
+      }
+    });
+  }
```

Comment:
I am not sure if this is the solution. Now, we have to remember to overwrite the `user` object before we use this.

The mutation is unwieldy. Either:
a) Do not require any of the previous properties we currently mutate, and, instead, keep this function simpler/purer
b) Pass in an `updatedFields` object containing the properties needing mutated: e.g. `updateUserChallengeData(fastify, user, challengeId, completedChallenge, timezone, updatedFields)`

Reply:
I believe that would be a post-MVP change. Also we need to remember to mutate the user object only for one submission endpoint.

Reply:
It's fine to refactor code during the migration. We have to ensure that the behaviour doesn't break the client or the mobile app (or the db).

That said, while I agree with you (Shaun) that this function is quite unwieldy, I don't mind using it (briefly) to help build these endpoints. As long as we refactor soon.

Reply:
https://github.com/freeCodeCamp/freeCodeCamp/blob/22bf130fd8498eb2385b5678c89b380b33036a2d/api/src/routes/helpers/challenge-helpers.ts#L19 is what I have in mind. From https://github.com/freeCodeCamp/freeCodeCamp/pull/50701

Reply:
Just to confirm... If I submit a RWD certification project, I now need moderation and the flag gets updated in the DB.

Does this mean that completing the challenge results in two DB writes - one for the flag, and one for the completion?

Reply:
> That said, while I agree with you (Shaun) that this function is quite unwieldy, I don't mind using it (briefly) to help build these endpoints. As long as we refactor soon.

I agree with this and prefer it this way as there's less chance of something breaking. As when the endpoint is created, its tests will also have been created and so theres more chances of catching bugs then.

Reply:
Very well, but are we going to address Naomi's concern about doing 3 user update ops in one function call?

Reply:
I think we agree on the end goal (streamlined, readable update functions that mean we only call the DB once). If that's the case, I see two options: 

1. add a bunch of TODOs (or track this in an issue, or both), merge this and then refactor before the MVP.
2. build the functions we want now

I like the second option (that's why I did it in my PR 🤷‍♂️), but option one is perfectly fine, too.

Reply:
Well, it is not used anywhere yet. So, provided we remember to not blindly use it, I like the option of merging this to carry on.

Reply:
Sounds good. Can you approve it?